## Trabajo final de Análisis de datos: Recorridos en Ecobici

Integrantes:
- Alejandro García Peña
- Juan Francisco Nadal
- Elian Ricardo Pinzas

Docentes:
- Esp. Lic. María Carina Roldán
- Esp. Ing. Ariadna Garmendia

Carrera de Especialización en Inteligencia Artificial  
Laboratorio de Sistemas Embebidos  
Facultad de Ingeniería de la Universidad de Buenos Aires  

### Introducción

- Para este trabajo práctico se eligió el set de datos __"Recorridos en Ecobici"__.
- __Ecobici__ o __“BA Ecobici”__ es un sistema de bicicletas compartidas que funciona en la Ciudad de Buenos Aires.
-  Comenzó a funcionar en 2010 con un total de 3 estaciones y 72 bicicletas.
- En la actualidad posee más de 380 estaciones y 3600 bicicletas con presencia en 47 de los 48 barrios porteños.
- Se realizan más de 200,000 viajes por mes
- Forma de uso: el sistema requiere que el usuario se registre y pague un pase. Una vez computado el pago, se permite disponer de la bicicleta por una determinada cantidad de tiempo. Al finalizar su uso, la bicicleta debe dejarse en un punto de anclaje dentro de alguna de las estaciones habilitadas.
- El servicio se encuentra actualmente concesionado a Tembici (empresa brasileña) desde 2019 y por 10 años.


### Introducción: set de datos

Dentro del repositorio de la Ciudad de Buenos Aires (https://data.buenosaires.gob.ar/dataset/bicicletas-publicas) pueden encontrarse distintos sets de datos correspondientes a distintos años.
- El año elegido para desarrollar el presente TP es 2024. Dicho archivo es _badata_ecobici_recorridos_realizados_2024.zip_ (161 MB). Para cada registro allí contenido se describe un viaje registrado desde una estación inicial hasta una estación destino indicando, adicionalmente, otros parámetros del viaje (por ejemplo: tiempo insumido, datos de la estación de origen y destino, coordenadas de las estaciones, sexo del usuario, etc)
- Adicionalmente en el repositorio puede encontrarse otro archivo correspondiente al año 2024: _usuarios_ecobici_2024.xls_. Ese set de datos indica los usuarios que se dieron de alta durante ese año.

A los fines del presente TP, el archivo de alta de usuarios será ignorado. El desarrollo se centrará en el archivo de recorridos (_badata_ecobici_recorridos_realizados_2024.zip_.)

# Consignas

## Carga del dataset

In [3]:
import pandas as pd

In [4]:
df_ecobici = pd.read_csv("./dataset/badata_ecobici_recorridos_realizados_2024.zip")

## Análisis exploratorio de los datos

In [5]:
df_ecobici.head()

,id_recorrido,duracion_recorrido,fecha_origen_recorrido,id_estacion_origen,nombre_estacion_origen,direccion_estacion_origen,long_estacion_origen,lat_estacion_origen,fecha_destino_recorrido,id_estacion_destino,nombre_estacion_destino,direccion_estacion_destino,long_estacion_destino,lat_estacion_destino,id_usuario,modelo_bicicleta,genero
0,20428222,568,2024-01-23 18:36:00,513,308 - SAN MARTIN II,Av. San Martín 5129,-58.490739,-34.597130,2024-01-23 18:45:28,498,055 - HABANA,Gral. José Gervasio Artigas 4298 (y Habana),-58.494959,-34.586598,992557.0,FIT,MALE
1,20431744,1355,2024-01-23 22:41:20,460,133 - BEIRO Y SEGUROLA,Segurola 3194,-58.511930,-34.607500,2024-01-23 23:03:55,382,204 - Biarritz,Biarritz 2403,-58.477255,-34.605431,320782.0,FIT,FEMALE
2,20429936,0,2024-01-23 20:06:22,467,328 - SARMIENTO II,Sarmiento 2037,-58.395893,-34.605514,2024-01-23 20:06:22,6,006 - Parque Lezama,"Avenida Martin Garcia, 295",-58.369758,-34.628526,828678.0,FIT,FEMALE
3,20429976,0,2024-01-23 20:08:17,382,204 - Biarritz,Biarritz 2403,-58.477255,-34.605431,2024-01-23 20:08:17,460,133 - BEIRO Y SEGUROLA,Segurola 3194,-58.511930,-34.607500,320782.0,ICONIC,FEMALE
4,20424802,680,2024-01-23 15:18:39,137,137 - AZOPARDO Y CHILE,AZOPARDO 700,-58.367492,-34.615598,2024-01-23 15:29:59,150,150 - RODRIGO BUENO,Av. España 2200,-58.355465,-34.618755,861425.0,FIT,FEMALE


In [6]:
df_ecobici.info()

<class 'pandas.DataFrame'>
RangeIndex: 3559284 entries, 0 to 3559283
Data columns (total 17 columns):
 #   Column                      Dtype  
---  ------                      -----  
 0   id_recorrido                int64  
 1   duracion_recorrido          int64  
 2   fecha_origen_recorrido      str    
 3   id_estacion_origen          int64  
 4   nombre_estacion_origen      str    
 5   direccion_estacion_origen   str    
 6   long_estacion_origen        float64
 7   lat_estacion_origen         float64
 8   fecha_destino_recorrido     str    
 9   id_estacion_destino         int64  
 10  nombre_estacion_destino     str    
 11  direccion_estacion_destino  str    
 12  long_estacion_destino       float64
 13  lat_estacion_destino        float64
 14  id_usuario                  float64
 15  modelo_bicicleta            str    
 16  genero                      str    
dtypes: float64(5), int64(4), str(8)
memory usage: 461.6 MB


### Características principales

- El set de datos posee 3559284 entradas
- Variables y tipos de datos:

| Nombre | Descripción | Tipo |
|-|-|-|
|id_recorrido                   | Número identificador de recorrido                    | Categórica nominal     |
| duracion_recorrido            | Duración del recorrido                               | Numérica               |
| fecha_origen_recorrido        | Fecha en que inició el recorrido                     | Categórica / Numérica  |
| id_estacion_origen            | Número identificador de estación de origen           | Categórica nominal     |
| nombre_estacion_origen        | Nombre de estación de origen                         | Categórica nominal     |
| direccion_estacion_origen     | Dirección de estación de origen                      | Categórica nominal     |
| long_estacion_origen          | Longitud de estación de origen                       | Numérica               |
| lat_estacion_origen           | Latitud de estación de origen                        | Numérica               |
| fecha_destino_recorrido       | Fecha en que finalizó el recorrido                   | Categórica / Numérica  |
| id_estacion_destino           | Número identificador de estación de destino          | Categórica nominal     |
| nombre_estacion_destino       | Nombre de estación de destino                        | Categórica nominal     |
| direccion_estacion_destino    | Dirección de estación de destino                     | Categórica nominal     |
| long_estacion_destino         | Longitud (coordenada) de estación de destino         | Numérica               |
| lat_estacion_destino          | Latitud (coordenada) de estación de destino          | Numérica               |
| id_usuario                    | Número identificador de usuario                      | Categórica nominal     |
| modelo_bicicleta              | Nombre del modelo de la bicicleta usada en recorrido | Categórica nominal     |
| genero                        | Género del usuario del recorrido                     | Categórica binaria     |

### Ajuste de variables

In [7]:
df_ecobici_numericas = ['duracion_recorrido', 'long_estacion_origen', 'lat_estacion_origen', 'long_estacion_destino', 'lat_estacion_destino']
df_ecobici_categoricas = ['id_recorrido', 'fecha_origen_recorrido', 'id_estacion_origen', 'nombre_estacion_origen', 'direccion_estacion_origen', 'fecha_destino_recorrido', 'id_estacion_destino', 'nombre_estacion_destino', 'direccion_estacion_destino', 'id_usuario', 'modelo_bicicleta', 'genero']

# Mapping para numericas y para categoricas
dtype_map = {col: 'float64'  for col in df_ecobici_numericas}
dtype_map.update({col: 'category' for col in df_ecobici_categoricas}) 

# Se aplica el mapping al dataframe
df_ecobici = df_ecobici.astype(dtype_map)

df_ecobici.dtypes

id_recorrido                  category
duracion_recorrido             float64
fecha_origen_recorrido        category
id_estacion_origen            category
nombre_estacion_origen        category
direccion_estacion_origen     category
long_estacion_origen           float64
lat_estacion_origen            float64
fecha_destino_recorrido       category
id_estacion_destino           category
nombre_estacion_destino       category
direccion_estacion_destino    category
long_estacion_destino          float64
lat_estacion_destino           float64
id_usuario                    category
modelo_bicicleta              category
genero                        category
dtype: object

### Estadística descriptiva

#### Variables numéricas

In [9]:
print("Duración de recorridos (en minutos)")
print(f"Media: {df_ecobici['duracion_recorrido'].mean() / 60}")
print(f"Mediana: {df_ecobici['duracion_recorrido'].median() / 60}")
print(f"Moda: {df_ecobici['duracion_recorrido'].mode() / 60}")

Duración de recorridos (en minutos)
Media: 21.24711279478307
Mediana: 14.683333333333334
Moda: 0    0.033333
Name: duracion_recorrido, dtype: float64


- Exploración y comprensión de los datos:
    - Cargar el dataset proporcionado y realizar un análisis exploratorio de los datos.
    - Describir las características principales del dataset, incluyendo el número de observaciones, número de variables y tipos de datos.
    - Identificar patrones generales y distribuciones.
    - Identificar errores, outliers (anomalías), valores faltantes y su tipo (MCAR, MAR, MNAR).
- Aplicación de técnicas de visualización:
    - Utilizar los gráficos y técnicas de visualización aprendidos para ilustrar las principales características del dataset.
    - Emplear el gráfico correcto en función del tipo de variable y del análisis que se está realizando.
    - Interpretar los resultados obtenidos a partir de las visualizaciones.
    - Plantear un posible problema de ML supervisado a partir de los datos elegidos.
    - Describir el problema de clasificación o de regresión.
    - Definir la variable target.
- Preprocesamiento y limpieza del dataset:
    - Realizar una limpieza general del dataset, eliminando o corrigiendo datos inconsistentes o irrelevantes.
    - Realizar el split del dataset (ej: train y test).
    - Identificar y tratar los valores faltantes en el dataset.
    - Detectar y manejar los outliers utilizando técnicas estadísticas o visuales apropiadas.
    - Escalar/normalizar los features.
- Feature engineering:
    - Crear nuevos features en caso de ser necesario. Justificar.
    - Aplicar técnicas de conversión de variables: codificación, discretización.
    - Analizar el balance/desbalance de clases (en el caso que se trate de un problema de clasificación).
    - Proponer y aplicar mecanismos de balance en caso de ser necesario y justificar la selección.
- Reducción de dimensionalidad
    - Evaluar relaciones entre variables y realizar una selección de features con los mecanismos vistos en clase (ej: filtros).
    - Implementar la técnica de extracción de features vista en clase (PCA). Evaluar ventajas y desventajas de la reducción.